lib/saluki-components/src/sources/dogstatsd/origin.rs

```diff
@@ -0,0 +1,304 @@
+    entity_id_precedence: bool,
+    tag_cardinality: OriginTagCardinality,
+    origin_detection_unified: bool,
+    origin_detection_optout: bool,
+    origin_cache: HashMap<OriginKey, SharedTagSet>,
+}
+
+impl<W> DogStatsDOriginEnricher<W>
+where
+    W: WorkloadProvider + Send + Sync,
+{
+    fn resolve_origin(&self, origin_info: &OriginInfo<'_>) -> Option<OriginKey> {
+        // Calculate the key for this origin information, and see if we've already previously resolved it.
+        let origin_key = OriginKey::from_opaque(origin_info);
+        if let Some(tags) = self.origin_cache.pin().get(&origin_key) {
+            trace!(origin = %origin_info, tags_len = tags.len(), "Found existing origin during resolving.");
+            return Some(origin_key);
+        }
+
+        // We couldn't find the origin information in the cache, so we'll try to resolve it now.
+        let mut had_entity_matches = false;
+        let mut enriched_tags = TagSet::default();
+
+        // Examine the various possible entity ID values, and based on their state, use one or more of them to grab any
+        // enriched tags attached to the entities. Below is a description of each entity ID we may have extracted:
+        //
+        // - entity ID (extracted from `dd.internal.entity_id` tag; non-prefixed pod UID)
+        // - container ID (extracted from `saluki.internal.container_id` tag, which comes from special "container ID"
+        //   extension in DogStatsD protocol; non-prefixed container ID)
+        // - origin PID (extracted via UDS socket credentials)
+        let maybe_entity_id = origin_info.pod_uid().and_then(EntityId::from_pod_uid);
+        let maybe_container_id = origin_info.container_id().and_then(EntityId::from_raw_container_id);
+        let maybe_origin_pid = origin_info.process_id().map(EntityId::ContainerPid);
+
+        let tag_cardinality = origin_info.cardinality().unwrap_or(self.tag_cardinality);
+
+        if !self.origin_detection_unified {
+            if self.origin_detection_optout && tag_cardinality == OriginTagCardinality::None {
+                trace!("Skipping origin enrichment for DogStatsD metric with cardinality 'none'.");
+                return None;
+            }
+
+            // If we discovered an entity ID via origin detection, and no client-provided entity ID was provided (or it was,
+            // but entity ID precedence is disabled), then try to get tags for the detected entity ID.
+            if let Some(origin_pid) = maybe_origin_pid {
+                if maybe_entity_id.is_none() || !self.entity_id_precedence {
+                    match self.workload_provider.get_tags_for_entity(&origin_pid, tag_cardinality) {
+                        Some(tags) => {
+                            trace!(entity_id = ?origin_pid, tags_len = tags.len(), "Found tags for entity.");
+
+                            had_entity_matches = true;
+                            enriched_tags.merge_missing_shared(&tags);
```

Comment:
This could be a later optimization, since I know we're already in the semi-slow path here, but it does seem like we don't necessarily need to materialize the full tagset here. If the workload provider just returned a stable reference to the origin tagset, we could use something like `ChainedTagSet` or more general structural sharing to mostly avoid allocating here. It should also reduce memory use by a meaningful amount, since we'd stop duplicating the origin tags in memory.